src/main/java/com/synopsys/integration/detect/tool/signaturescanner/BlackDuckSignatureScannerTool.java

```diff
@@ -63,71 +63,90 @@ public BlackDuckSignatureScannerTool(BlackDuckSignatureScannerOptions signatureS
+        ScanBatchRunner scanBatchRunner = createScanBatchRunner(blackDuckServerConfig, localScannerInstallPath.orElse(null), scanBatchRunnerFactory, installDirectory, connectionFactory);
+
+        try {
+            BlackDuckSignatureScanner blackDuckSignatureScanner = detectContext.getBean(BlackDuckSignatureScanner.class, signatureScannerOptions, scanBatchRunner, blackDuckServerConfig, codeLocationNameManager);
+            return runScanner(blackDuckSignatureScanner, codeLocationCreationService, blackDuckServerConfig, projectNameVersion, installDirectory, dockerTar.orElse(null));
+        } catch (IOException | IntegrationException e) {
+            logger.error(String.format("Signature scan failed: %s", e.getMessage()));
+            logger.debug("Signature scan error", e);
+            return SignatureScannerToolResult.createFailureResult();
+        } finally {
+            executorService.shutdownNow();
+        }
+    }
+
+    private Optional<Path> determineLocalScannerInstallPath() {
         Optional<Path> localScannerInstallPath = signatureScannerOptions.getOfflineLocalScannerInstallPath();
         if (!localScannerInstallPath.isPresent()) {
             localScannerInstallPath = signatureScannerOptions.getOnlineLocalScannerInstallPath();
         }
         localScannerInstallPath.ifPresent(path -> logger.debug(String.format("Determined local scanner path: %s", path.toString())));
+        return localScannerInstallPath;
+    }
 
-        BlackDuckSignatureScannerOptions blackDuckSignatureScannerOptions = detectConfigurationFactory.createBlackDuckSignatureScannerOptions();
-        ExecutorService executorService = Executors.newFixedThreadPool(blackDuckSignatureScannerOptions.getParallelProcessors());
-        IntEnvironmentVariables intEnvironmentVariables = IntEnvironmentVariables.includeSystemEnv();
+    private File determineInstallDirectory(@Nullable Path localScannerInstallPath, DirectoryManager directoryManager) {
+        logger.debug("Determining signature scanner install directory...");
+        File installDirectory = directoryManager.getPermanentDirectory();
 
-        ScanBatchRunnerFactory scanBatchRunnerFactory = new ScanBatchRunnerFactory(intEnvironmentVariables, executorService);
+        if (localScannerInstallPath != null) {
+            logger.debug(String.format("Using provided path: %s", localScannerInstallPath));
+            installDirectory = localScannerInstallPath.toFile();
+        } else {
+            logger.debug("Using default scanner path.");
+        }
+        logger.debug(String.format("Determined install directory: %s", installDirectory.getAbsolutePath()));
+        return installDirectory;
+    }
+
+    private ScanBatchRunner createScanBatchRunner(@Nullable BlackDuckServerConfig blackDuckServerConfig, @Nullable Path localScannerInstallPath, ScanBatchRunnerFactory scanBatchRunnerFactory, File installDirectory, ConnectionFactory connectionFactory) throws DetectUserFriendlyException {
```

Comment:
Seems like this class has multiple responsibilities-- creating the ScanBatchRunner and then running the signature scanner. Can we move creating the ScanBatchRunner into another class and then demand the scanBatchRunner and the installDirectory in this of this one?
determineLocalScannerInstallPath and determineInstallDirectory also appear that they could move with it.

Reply:
If you did this, I suspect you could move RunScanner back into runScanTool, since that would be this class' only responsibility.